internal/db/schema/manager.go

```diff
@@ -191,9 +193,20 @@ func (b *Manager) runMigrations(ctx context.Context, qp *statementProvider) erro
 		default:
 			// context is not done yet. Continue on to the next query to execute.
 		}
-		if err := b.driver.Run(ctx, bytes.NewReader(qp.ReadUp()), qp.Version()); err != nil {
+		upVer := qp.ReadUp()
+		if upVer.PreHook != nil {
```

Comment:
Instead of Pre/Post hooks, what if we decoupled them completely from a statement version?  What I am envisioning would be each version could be either an SQL statement that is passed to driver.Run(ctx, ...) or a goFunction that is passed to something like driver.ExecuteFn(ctx, migrateFunction) but not both.

The benefit of this approach, in my opinion, is that it allows the sql.Tx to not have to be exported since there may be a db in the future that we rely on dirty bits instead of transactions.  It also makes the migration steps a little easier to reason about since you can just step through each migration version sequentially.

Reply:
So to make sure I'm understanding, you are suggesting that rather than pass statements, that we encapsulate the entirety of a version in a Go function? That way a "basic" update function would just execute the statements, but then rather than pre/post hooks we can just mix in logic as needed?